### ISSUE TITLE
NIST CAVP validation of liboqs ML-DSA algorithm

### ISSUE THREAD

**Issue by stauro79** [open-quantum-safe/liboqs#1931]
Is there any road-map/plan for liboqs ML-DSA algorithm to be certified using NIST Cryptographic Algorithm Validation Program (CAVP)?

**Comment by SWilson4** [comment]
Thanks for the question, @stauro79. I'm not aware of any plans to do this, though perhaps it's something we could do after the final standard is integrated (in 0.12.0). That said, I imagine we'd want to have a very stable implementation of any algorithm we have certified: perhaps this would be a good criterion for a [future 1.0.0 release](https://github.com/orgs/open-quantum-safe/discussions/1892)?

Assuming you're a (possibly potential) consumer of liboqs, is CAVP validation something that you would find valuable and/or essential?

**Comment by stauro79** [comment]
Thanks @SWilson4 .  Yes it is essential as there is a requirement to use/deploy CAVP validated cryptographic algorithms

**Comment by dstebila** [comment]
I believe @ashman-p had been looking at the CAVP test vectors.

**Comment by bhess** [comment]
ML-KEM in liboqs is currently tested against NIST's static [ACVP vectors](https://github.com/open-quantum-safe/liboqs/blob/main/tests/test_acvp_vectors.py). Planning to do the same for the ML-DSA integration. AFAIK the same type of tests are performed during a certification.

**Comment by baentsch** [comment]
> ML-KEM in liboqs is currently tested against NIST's static [ACVP vectors](https://github.com/open-quantum-safe/liboqs/blob/main/tests/test_acvp_vectors.py). Planning to do the same for the ML-DSA integration. AFAIK the same type of tests are performed during a certification.

Great! Thanks for reminding (at least me). And the reason we're not landing https://github.com/open-quantum-safe/liboqs/pull/1919 is because we're waiting on APIs to become available enabling this test, right? Just for my personal curiosity now: There's no such additional API needed for MLKEM? And to the rest of the team: What are our plans wrt SLH-DSA in this regard? Worth while adding to https://github.com/open-quantum-safe/liboqs/issues/1894?

**Comment by stauro79** [comment]
> * This arguably should not just apply to ML-DSA but all newly standardized PQ algs.
> * What is the "cryptographic module" in CAVP parlance that you'd want to use/see CAVP validated? `liboqs` straight or rather any of its integrations (openssl, openssl, etc.), @stauro79 ?

Thanks @baentsch.  yes, validation would apply to all the PQC algs defined in the new NIST standard.   I am using liboqs through OpenSSL OQS Provider module.

**Comment by bhess** [comment]
> And the reason we're not landing #1919 is because we're waiting on APIs to become available enabling this test, right? Just for my personal curiosity now: There's no such additional API needed for MLKEM?

Yes to both. Plus to incorporate the tests against the external API that NIST announced for [around October](https://github.com/open-quantum-safe/liboqs/issues/1891#issuecomment-2320297333).

**Comment by baentsch** [comment]
Thanks @bhess for the confirmation(s).

@stauro79 this comment lets me wonder though:

> I am using liboqs through OpenSSL OQS Provider module.

Wouldn't you then need to have "certification" applied at that module level instead as it also/further "twiddles around" with the algorithms (not the core logic, though)? Or even better, at OpenSSL level itself (further "twiddling" there :)?

**Comment by stauro79** [comment]
The openssl has released OpenSSL FIPS Provider (crypto module) that is certificated by NIST.   https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4282  

Incase of liboqs and oqs provider, I am not sure how certification should work..  liboqs is crypto library and oqs provider just consumes the liboqs crypto implementation.

**Comment by baentsch** [comment]
> Incase of liboqs and oqs provider, I am not sure how certification should work.. liboqs is crypto library and oqs provider just consumes the liboqs crypto implementation.

Personally, I do not think there is the slightest chance this software combination can ever be getting FIPS certified. I think the best chance for that is getting [ML-DSA integrated directly into OpenSSL](https://github.com/openssl/openssl/issues/25961) (and there, with all required prerequisites (that are unknown to me) into the FIPS provider).

**Comment by SWilson4** [comment]
Our ML-DSA code is already tested against ACVP test vectors: see also https://github.com/open-quantum-safe/liboqs/issues/2061.
